docs/design_spec.md

```diff
@@ -0,0 +1,94 @@
+
+[![Nimbess Pipelines](./img/Nimbess_Pipelines.png)](./img/Nimbess_Pipelines.png)
+
+In the example above, the EgressPort for Port 1 is attached to every other Port Pipeline (in this case only 1)
+other than Port 1's Pipeline. The same is true for Port 2. This behavior prevents forwarding packets out the
+same interface traffic came in on. Each instance of the Meta Pipeline for the Port Pipeline uses a unique ID
+in order to distinguish these functions when they are rendered into different data planes. There may be many
+Meta Pipelines which belong to either different L2 networks (i.e. VLANs) or L3 networks (i.e. VRFs). A Port
+may only belong to a single Meta Pipeline.
+
+## Southbound Design
+
+Nimbess Agent uses a data plane agnostic southbound. Nimbess defines an abstract Driver type which any
+data plane driver may implement. The driver is configured in the Nimbess Agent config file and loaded at
+run time. A data plane driver is responsible for taking in Port Pipeline objects translating them into driver
+specific configuration, and then updating the Port Pipeline Modules with any related network configuration.
+This returned network configuration may vary per module, but as an example, updating a new Port object with
+the MAC address returned from the data plane when the Port is created.
+
+## Example Code Flow - CNI ADD Request
+
+Below is an example of walking through the pseudo code of a CNI ADD request:
+
+[![CNI CodeFlow](./img/code2flow_IvSpoQ.svg)](./img/code2flow_IvSpoQ.svg)
+
+A Nimbess Pipeline structure is used to hold all of the information needed to define a network pipeline. It includes
+information such as an array of all modules in the pipeline, the MetaKey for this pipeline, which port this
+pipeline belongs to, etc. While the pipeline holds the information describing which modules belong to the
+pipeline, it does not indicate module connectivity. Connectivity information is conveyed by the module itself,
+via Ingress and Egress Port Maps which contain a map of the gate number and a pointer to the next module this module is
+connected to, creating a multi-dimensional linked list. Therefore when a new Port Pipeline is created, it is necessary
+to make a copy of the Meta Pipeline, including a list of the Meta Modules used to instantiate a new Port Pipeline
+module list. However after creating the new Module list, all of the connectivity needs to be updated to point to the
+new instances of modules held within the Gate Maps. This is done via the above block which describes traversing the
+module linked list and updating it.
+
+After the new Port Pipeline is created, Ingress Port is created and attached to the Port Pipeline, and then rendered
+into the data plane via a driver call. Upon success of programming the data plane with the initial pipeline
+a lookup is performed to find other already existing Egress Ports which belong to this same network (MetaKey). If any
```

Comment:
should egress ports be added earlier? Wondering if the pipeline is installed and traffic hits it would it just be dropped if egress ports are not there.

Reply:
maybe could make it a single driver call to render all of it at once. optimization for the future.